app/controllers/picture_controller.rb

```diff
@@ -0,0 +1,33 @@
+class PictureController < ApplicationController
+  def show # GET /pictures/:basename
+    compressed_id, extension = params[:basename].split('.')
+    picture = Picture.find_by_id(Picture.uncompress_id(compressed_id))
+    if picture.blank? || extension != picture.extension
+      render :nothing => true, :status => 404
+    else
+      render_picture_content(picture)
+    end
+  end
+
+  private
+
+  def render_picture_content(picture)
+    response.headers['Cache-Control'] = "public"
+    response.headers['Content-Type'] = "image/#{picture.extension}"
+    response.headers['Content-Disposition'] = "inline"
+    render :text => cached_picture_content(picture)
+  end
+
+  def cached_picture_content(picture)
```

Comment:
performance, size and md5 are calculated with blob writes, fast to check against cache before reading picture content.